File: Algorithms/cpp/merge_sort.cpp
/**
========================================================================
@file merge_sort.cpp

Big O:

Space Complexity: O(n)
    - New arrays are required, however they take the same space in memory.

Time Complexity: 
    = O(n log(n) + n)
    = O(n log(n))       // drop non dominants.


: zach wolpe
: zach.wolpe@medibio.com.au
: 14 July 2023
========================================================================
*/


#include <iostream>

void merge(int array[], int leftIndex, int midIndex, int rightIndex) {
    
    // split array into two
    int leftArraySize  = midIndex - leftIndex + 1;
    int rightArraySize = rightIndex - midIndex;
    int leftArray[leftArraySize];
    int rightArray[rightArraySize];
    for (int i=0; i<leftArraySize; i++)
        leftArray[i] = array[leftIndex + i];
    for (int i=0; i<rightArraySize; i++)
        rightArray[i] = array[midIndex + i + 1];

    // sort by selecting the minimum at each index
    int index = leftIndex;
    int i = 0;
    int j = 0;

    while (i < leftArraySize && j < rightArraySize)
        if (leftArray[i] <= rightArray[j]) {
            array[index] = leftArray[i];
            index++;
            i++;
        } else {
            array[index] = rightArray[j];
            index++;
            j++;
        }
    
    // complete cycle
    while (i < leftArraySize) {
        array[index] = leftArray[i];
        index++;
        i++;
    };
    while (j < rightArraySize) {
        array[index] = rightArray[j];
        index++;
        j++;
    };

};

void mergeSort(int array[], int leftIndex, int rightIndex) {
    if (leftIndex >= rightIndex) return;
    int midIndex = leftIndex + (rightIndex - leftIndex) / 2;
    mergeSort(array, leftIndex,  midIndex);
    mergeSort(array, midIndex+1, rightIndex);
    merge(array, leftIndex, midIndex, rightIndex);

};

int main() {
    int myArray[] = {1,3,7,8,45,3,4,2,4,5,6};
    int size = sizeof(myArray) / sizeof(myArray[0]);
    int leftIndex = 0;
    int rightIndex = size - 1;
    mergeSort(myArray, leftIndex, rightIndex);
    for (auto value : myArray) {
        std::cout << value << " ";
    }
    return 0;

};

File: Algorithms/python/merge_sort.py
"""
========================================================================
@file merge_sory.py

Big O:

Space Complexity: O(n)
    - New arrays are required, however they take the same space in memory.

Time Complexity: 
    = O(n log(n) + n)
    = O(n log(n))       // drop non dominants.


: zach wolpe
: zach.wolpe@medibio.com.au
: 18 July 2023
========================================================================
"""

def merge(arr1, arr2):
    combined = []
    i = 0
    j = 0
    while i < len(arr1) and j < len(arr2):
        if arr2[j] > arr1[i]:
            combined.append(arr1[i])
            i += 1
        else:
            combined.append(arr2[j])
            j += 1
    
    while i < len(arr1):
        combined.append(arr1[i])
        i += 1
    
    while j < len(arr2):
        combined.append(arr2[j])
        j += 1

    return combined


def merge_sort(my_list):
    if len(my_list) == 1:
        return my_list
    mid = int(len(my_list)/2)
    left = my_list[:mid]
    right = my_list[mid:]
    return merge(merge_sort(left), merge_sort(right))




if __name__ == '__main__':
    
    array = [1, 222, 233, 4, 5, 6, 97, 8, 9]
    array = merge_sort(array)
    [print(i, end=' -> ') for i in array]
    print()